File: src/main/java/org/tasks/reminders/SnoozeActivity.java
package org.tasks.reminders;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.FragmentManager;

import com.todoroo.astrid.dao.TaskDao;
import com.todoroo.astrid.data.Task;

import org.tasks.activities.DateAndTimePickerActivity;
import org.tasks.activities.TimePickerActivity;
import org.tasks.injection.ActivityComponent;
import org.tasks.injection.InjectingAppCompatActivity;
import org.tasks.notifications.NotificationManager;
import org.tasks.time.DateTime;

import javax.inject.Inject;

public class SnoozeActivity extends InjectingAppCompatActivity implements SnoozeCallback, DialogInterface.OnCancelListener {

    private static final String FRAG_TAG_SNOOZE_DIALOG = "frag_tag_snooze_dialog";
    private static final String EXTRA_PICKING_DATE_TIME = "extra_picking_date_time";
    private static final int REQUEST_DATE_TIME = 10101;

    public static final String EXTRA_TASK_ID = "id";
    public static final String EXTRA_SNOOZE_TIME = "snooze_time";

    @Inject NotificationManager notificationManager;
    @Inject TaskDao taskDao;

    private long taskId;
    private boolean pickingDateTime;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setup(getIntent(), savedInstanceState);
    }

    @Override
    public void inject(ActivityComponent component) {
        component.inject(this);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);

        setup(intent, null);
    }

    private void setup(Intent intent, Bundle savedInstanceState) {
        taskId = intent.getLongExtra(EXTRA_TASK_ID, 0L);

        if (savedInstanceState != null) {
            pickingDateTime = savedInstanceState.getBoolean(EXTRA_PICKING_DATE_TIME, false);
            if (pickingDateTime) {
                return;
            }
        }

        if (intent.hasExtra(EXTRA_SNOOZE_TIME)) {
            snoozeForTime(new DateTime(intent.getLongExtra(EXTRA_SNOOZE_TIME, 0L)));
        } else {
            FragmentManager fragmentManager = getSupportFragmentManager();
            SnoozeDialog fragmentByTag = (SnoozeDialog) fragmentManager.findFragmentByTag(FRAG_TAG_SNOOZE_DIALOG);
            if (fragmentByTag == null) {
                fragmentByTag = new SnoozeDialog();
                fragmentByTag.show(fragmentManager, FRAG_TAG_SNOOZE_DIALOG);
            }
            fragmentByTag.setOnCancelListener(this);
            fragmentByTag.setSnoozeCallback(this);
        }
    }

    @Override
    public void snoozeForTime(DateTime time) {
        Task task = new Task();
        task.setId(taskId);
        task.setReminderSnooze(time.getMillis());
        taskDao.save(task);
        notificationManager.cancel(taskId);
        setResult(RESULT_OK);
        finish();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean(EXTRA_PICKING_DATE_TIME, pickingDateTime);
    }

    @Override
    public void pickDateTime() {
        pickingDateTime = true;

        Intent intent = new Intent(this, DateAndTimePickerActivity.class);
        intent.putExtra(DateAndTimePickerActivity.EXTRA_TIMESTAMP, new DateTime().plusMinutes(30).getMillis());
        startActivityForResult(intent, REQUEST_DATE_TIME);
    }

    @Override
    public void onCancel(DialogInterface dialog) {
        finish();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_DATE_TIME) {
            if (resultCode == RESULT_OK && data != null) {
                long timestamp = data.getLongExtra(TimePickerActivity.EXTRA_TIMESTAMP, 0L);
                snoozeForTime(new DateTime(timestamp));
            } else {
                finish();
            }
        } else {
            super.onActivityResult(requestCode, resultCode, data);
        }
    }
}

File: app/src/main/java/org/tasks/reminders/SnoozeActivity.kt
package org.tasks.reminders

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.activity.compose.setContent
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.lifecycle.lifecycleScope
import com.todoroo.astrid.alarms.AlarmService
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.launch
import org.tasks.R
import org.tasks.activities.DateAndTimePickerActivity
import org.tasks.analytics.Firebase
import org.tasks.compose.pickers.SnoozeDialog
import org.tasks.extensions.Context.is24HourFormat
import org.tasks.preferences.Preferences
import org.tasks.themes.TasksTheme
import org.tasks.themes.Theme
import org.tasks.time.DateTimeUtils2.currentTimeMillis
import org.tasks.time.ONE_MINUTE
import javax.inject.Inject

@AndroidEntryPoint
class SnoozeActivity : AppCompatActivity() {
    @Inject lateinit var alarmService: AlarmService
    @Inject lateinit var firebase: Firebase
    @Inject lateinit var preferences: Preferences
    @Inject lateinit var theme: Theme

    private val taskIds: MutableList<Long> = ArrayList()
    private var pickingDateTime = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val intent = intent
        if (intent.hasExtra(EXTRA_TASK_ID)) {
            taskIds.add(intent.getLongExtra(EXTRA_TASK_ID, -1L))
        } else if (intent.hasExtra(EXTRA_TASK_IDS)) {
            taskIds.addAll(intent.getSerializableExtra(EXTRA_TASK_IDS) as ArrayList<Long>)
        }
        if (savedInstanceState != null) {
            pickingDateTime = savedInstanceState.getBoolean(EXTRA_PICKING_DATE_TIME, false)
            if (pickingDateTime) {
                return
            }
        }
        if (intent.hasExtra(EXTRA_SNOOZE_TIME)) {
            snoozeForTime(intent.getLongExtra(EXTRA_SNOOZE_TIME, 0L))
            return
        }
        setContent {
            TasksTheme(
                theme = theme.themeBase.index,
                primary = theme.themeColor.primaryColor,
            ) {
                var visible by remember { mutableStateOf(true) }
                SnoozeDialog(
                    visible = visible,
                    loadTimes = { preferences.quickPickTimes },
                    is24Hour = is24HourFormat,
                    onSelected = { snoozeForTime(it) },
                    onPickDateTime = {
                        visible = false
                        pickDateTime()
                    },
                    onDismiss = { finish() },
                )
            }
        }
    }

    private fun snoozeForTime(time: Long) {
        firebase.logEvent(R.string.event_notification, R.string.param_type to "snooze_time")
        lifecycleScope.launch(NonCancellable) {
            alarmService.snooze(time, taskIds)
        }
        setResult(Activity.RESULT_OK)
        finish()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(EXTRA_PICKING_DATE_TIME, pickingDateTime)
    }

    private fun pickDateTime() {
        pickingDateTime = true
        val intent = Intent(this, DateAndTimePickerActivity::class.java)
        intent.putExtra(
            DateAndTimePickerActivity.EXTRA_TIMESTAMP,
            currentTimeMillis() + PICKER_OFFSET,
        )
        startActivityForResult(intent, REQUEST_DATE_TIME)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_DATE_TIME) {
            if (resultCode == Activity.RESULT_OK && data != null) {
                snoozeForTime(data.getLongExtra(DateAndTimePickerActivity.EXTRA_TIMESTAMP, 0L))
            } else {
                finish()
            }
        } else {
            super.onActivityResult(requestCode, resultCode, data)
        }
    }

    companion object {
        private const val FLAGS = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        const val EXTRA_TASK_ID = "id"
        const val EXTRA_TASK_IDS = "ids"
        const val EXTRA_SNOOZE_TIME = "snooze_time"
        private const val EXTRA_PICKING_DATE_TIME = "extra_picking_date_time"
        private const val REQUEST_DATE_TIME = 10101

        private val PICKER_OFFSET = 30 * ONE_MINUTE

        fun newIntent(context: Context?, id: Long?): Intent =
                Intent(context, SnoozeActivity::class.java).apply {
                    flags = FLAGS
                    putExtra(EXTRA_TASK_ID, id)
                }

        fun newIntent(context: Context?, ids: List<Long?>?): Intent =
                Intent(context, SnoozeActivity::class.java).apply {
                    flags = FLAGS
                    putExtra(EXTRA_TASK_IDS, ArrayList<Any?>(ids!!))
                }
    }
}
